crates/ide-assists/src/handlers/wrap_return_type_in_result.rs

```diff
@@ -1,16 +1,77 @@
 use std::iter;
 
+use hir::HasSource;
 use ide_db::{
     famous_defs::FamousDefs,
     syntax_helpers::node_ext::{for_each_tail_expr, walk_expr},
 };
+use itertools::Itertools;
 use syntax::{
-    ast::{self, make, Expr},
-    match_ast, ted, AstNode,
+    ast::{self, make, Expr, HasGenericParams},
+    match_ast, ted, AstNode, ToSmolStr,
 };
 
 use crate::{AssistContext, AssistId, AssistKind, Assists};
 
+fn result_ty(ctx: &AssistContext<'_>, ret_type: &ast::Type) -> Option<ast::Type> {
+    let scope = ctx.sema.scope(ret_type.syntax())?;
+    let core_result = FamousDefs(&ctx.sema, scope.krate()).core_result_Result()?;
+
+    let ty = ctx.sema.resolve_type(ret_type)?.as_adt();
+    if matches!(ty, Some(hir::Adt::Enum(ret_type)) if ret_type == core_result) {
+        // The return type is already wrapped in a Result
+        cov_mark::hit!(wrap_return_type_in_result_simple_return_type_already_result);
+        return None;
+    }
+
+    // Try to find a Result<T, ...> type alias in the current scope (shadowing the default).
+    let result_name = core_result.name(ctx.db());
+    let mut local_result = None;
+    scope.process_all_names(&mut |name, def| match def {
+        hir::ScopeDef::ModuleDef(hir::ModuleDef::TypeAlias(alias)) if name == result_name => {
+            let Some(enum_ty) = alias.ty(ctx.db()).as_adt().and_then(|adt| adt.as_enum()) else {
+                return;
+            };
+
+            let ret_is_alias = matches!(ty, Some(hir::Adt::Enum(ret_type)) if ret_type == enum_ty);
+            if enum_ty != core_result || ret_is_alias {
+                // The current name is not a Result alias, or the return type is already this Result alias.
+                return;
+            }
+
+            let generic_params = alias.source(ctx.db()).and_then(|src| {
+                let mut inserted_ret_type = false;
+                let generics = src
+                    .value
+                    .generic_param_list()?
+                    .generic_params()
+                    .map(|param| match param {
+                        // Replace the very first type parameter with the functions return type.
+                        ast::GenericParam::TypeParam(_) if !inserted_ret_type => {
+                            inserted_ret_type = true;
+                            ret_type.to_smolstr()
+                        }
+                        ast::GenericParam::LifetimeParam(_) => make::lifetime("'_").to_smolstr(),
+                        _ => make::ty_placeholder().to_smolstr(),
+                    })
+                    .join(", ");
+                inserted_ret_type.then_some(generics)
+            });
+            if let Some(generic_params) = generic_params {
+                let name = name.as_str();
+                local_result = Some(make::ty(&format!("{name}<{generic_params}>")));
+            }
+        }
```

Comment:
all of this processing can be done in the builder closure I think (which makes things a bit more lazy)